crates/config/src/compilation.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{filter::GlobMatcher, serde_helpers};
+use foundry_compilers::{
+    artifacts::{BytecodeHash, EvmVersion},
+    multi::{MultiCompilerRestrictions, MultiCompilerSettings},
+    settings::VyperRestrictions,
+    solc::{Restriction, SolcRestrictions},
+    RestrictionsWithVersion,
+};
+use semver::VersionReq;
+use serde::{Deserialize, Serialize};
+
+/// Keeps possible overrides for default settings which users may configure to construct additional
+/// settings profile.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SettingsOverrides {
+    pub name: String,
+    via_ir: Option<bool>,
+    #[serde(default, with = "serde_helpers::display_from_str_opt")]
+    evm_version: Option<EvmVersion>,
+    optimizer: Option<bool>,
+    optimizer_runs: Option<usize>,
+    bytecode_hash: Option<BytecodeHash>,
+}
+
+impl SettingsOverrides {
+    /// Applies the overrides to the given settings.
+    pub fn apply(&self, settings: &mut MultiCompilerSettings) {
+        if let Some(via_ir) = self.via_ir {
+            settings.solc.via_ir = Some(via_ir);
+        }
+
+        if let Some(evm_version) = self.evm_version {
+            settings.solc.evm_version = Some(evm_version);
+            settings.vyper.evm_version = Some(evm_version);
+        }
+
+        if let Some(enabled) = self.optimizer {
+            settings.solc.optimizer.enabled = Some(enabled);
+        }
+
+        if let Some(optimizer_runs) = self.optimizer_runs {
+            settings.solc.optimizer.runs = Some(optimizer_runs);
+        }
+
+        if let Some(bytecode_hash) = self.bytecode_hash {
+            if let Some(metadata) = settings.solc.metadata.as_mut() {
+                metadata.bytecode_hash = Some(bytecode_hash);
+            } else {
+                settings.solc.metadata = Some(bytecode_hash.into());
+            }
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RestrictionsError {
+    #[error("specified both exact and relative restrictions for {0}")]
+    BothExactAndRelative(&'static str),
+}
+
+/// Restrictions for compilation of given paths.
+///
+/// Only purpose of this type is to accept user input to later construct
+/// `RestrictionsWithVersion<MultiCompilerRestrictions>`.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CompilationRestrictions {
+    pub paths: GlobMatcher,
+    version: Option<VersionReq>,
+    via_ir: Option<bool>,
+    bytecode_hash: Option<BytecodeHash>,
+
+    min_optimizer_runs: Option<usize>,
+    optimizer_runs: Option<usize>,
+    max_optimizer_runs: Option<usize>,
+
+    #[serde(default, with = "serde_helpers::display_from_str_opt")]
```

Comment:
unrelated,
worth checking which of the serde_utils we'd like to have in alloy-serde